spectrum-network/src/protocol_handler/multicasting.rs

```diff
@@ -12,43 +12,65 @@ use libp2p_identity::PeerId;
 
 use algebra_core::{CommutativePartialSemigroup, CommutativeSemigroup};
 use spectrum_crypto::{AsyncVerifiable, VerifiableAgainst, Verified};
+use tracing::trace;
 
 use crate::protocol_handler::multicasting::overlay::DagOverlay;
 use crate::protocol_handler::pool::{FromTask, TaskPool};
 use crate::protocol_handler::void::VoidMessage;
 use crate::protocol_handler::{NetworkAction, ProtocolBehaviourOut, TemporalProtocolStage};
 
-use super::handel::partitioning::PeerIx;
+use super::handel::partitioning::{PeerIx, PeerPartitions};
 use super::handel::Weighted;
 
 pub mod overlay;
 
 /// DAG based multicasting that accumulates received statements along the way.
-pub struct DagMulticasting<S, P> {
+pub struct DagMulticasting<S, P, PP> {
     pub statement: Option<S>,
     pub public_data: P,
     pub overlay: DagOverlay,
     pub contacted_peers: HashSet<PeerId>,
     pub outbox: VecDeque<ProtocolBehaviourOut<VoidMessage, S>>,
+    partitions: PP,
     creation_time: std::time::Instant,
     processing_delay: Duration,
     next_processing: Option<Pin<Box<tokio::time::Sleep>>>,
     multicasting_duration: Duration,
 }
 
-impl<S, P> DagMulticasting<S, P> {
+impl<S, P, PP> DagMulticasting<S, P, PP>
+where
+    PP: PeerPartitions + Send + Clone,
+{
     pub fn new(
         statement: Option<S>,
         public_data: P,
         overlay: DagOverlay,
         config: DagMulticastingConfig,
+        partitions: PP,
```

Comment:
Is this only for tracing?

Reply:
It's to see in general what is being sent/received by a node for broadcasts. Was helpful for debugging.

Reply:
PeerIx values are much easier to trace than PeerId, which is what multicasting only had originally.